torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -143,66 +144,129 @@ class ProcessGroupNCCL : public ProcessGroup {
+    // ensure we can have async CUDA execution and it does not wait for the
+    // entire operation to complete on GPU.
     void wait() override {
-      work_->wait();
+      auto currentStream = at::cuda::getCurrentCUDAStream(deviceIndex_);
+      (*cudaEvent_).block(currentStream);
     }
 
-    // FutureNCCL's value is NCCL collective's outputs and callbacks were
-    // invoked inline by addCallback(), so markCompleted is not needed.
+    // FutureNCCL's value is NCCL collective's outputs or callback() output with
+    // callbacks being invoked inline by addCallbackWithStream(), so
+    // markCompleted is not needed.
     void markCompleted(at::IValue /* unused */) override {
       C10_THROW_ERROR(Error, "FutureNCCL::markCompleted is not supported.");
     }
 
-    // Just returns NCCL collective's outputs after WorkNCCL's wait returns.
+    // Just returns FutureNCCL's value.
     at::IValue value() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
     const at::IValue& constValue() override {
-      work_->wait();
-      return outputs_;
+      return value_;
     }
 
-    // Add a callback to FutureNCCL. It invokes the callback inline after
-    // WorkNCCL's wait(). workCallbacks return a Future (not FutureNCCL).
-    void addCallback(std::function<void(void)> callback) override {
-      work_->wait();
+    // Adds a callback to FutureNCCL. It invokes the callback inline after
+    // synchronizing FutureNCCL's own cudaEvent with the stream that runs
+    // this callback. FutureNCCL then will return a FutureNCCL if this callback
+    // is executed without any errors. This new FutureNCCL's cudaEvent will
+    // record the callback's stream.
+    void addCallbackWithStream(
+        std::function<void(void)> callback,
+        const c10::cuda::CUDAStream stream,
+        std::shared_ptr<at::cuda::CUDAEvent> thenFutCudaEvent) {
+      (*cudaEvent_).block(stream);
+      c10::OptionalStreamGuard streamGuard{c10::Stream(stream)};
       callback();
+      (*thenFutCudaEvent).record(stream);
+    }
+
+    // We use addCallbackWithStream instead of addCallback.
+    void addCallback(std::function<void(void)> /* unused */) override {
+      C10_THROW_ERROR(
+          Error,
+          "FutureNCCL uses addCallbackWithStream instead of addCallback.");
+    }
+
+    // Adds a callback to the futureNCCL, and returns another FutureNCCL to hold
+    // the return value of the callback and new cudaEvent that recorded the
+    // stream that runs this callback.
+    c10::intrusive_ptr<Future> then(
+        std::function<at::IValue(void)> callback,
+        at::TypePtr type) override {
+      // Set a default fut, this will be returned in case of an error in the
+      // callback. If the callback is executed without any errors, fut will be
+      // replaced by a FutureNCCL.
+      auto fut = c10::make_intrusive<Future>(type);
```

Comment:
This is a little messy since you're creating a Future of default type and then replacing it with FutureNCCL inside the callback. You probably want to create `FutureNCCL` here and then call `markCompleted()` inside the lambda below. For this, you need another `FutureNCCL` constructor that doesn't take `value_` as input. Then, `markCompleted()` allows the user to set the value only if it is not set already. You can use `value_->isNone()` to check if it has been set or not.